pallets/communities/src/functions.rs

```diff
@@ -122,76 +123,89 @@ impl<T: Config> Pallet<T> {
+			Self::update_locks(who)
 		})
 	}
 
-	pub(crate) fn do_remove_vote(
+	pub(crate) fn try_remove_vote(
 		who: &AccountIdOf<T>,
 		membership_id: MembershipIdOf<T>,
 		poll_index: PollIndexOf<T>,
 	) -> DispatchResult {
-		let Some(community_id) = T::MemberMgmt::check_membership(who, &membership_id) else {
-			fail!(Error::<T>::NotAMember);
-		};
+		let community_id = T::MemberMgmt::check_membership(who, &membership_id).ok_or(Error::<T>::NotAMember)?;
 
 		T::Polls::try_access_poll(poll_index, |poll_status| {
 			let Some((tally, class)) = poll_status.ensure_ongoing() else {
 				fail!(Error::<T>::NotOngoing);
 			};
 			ensure!(community_id == class, Error::<T>::InvalidTrack);
-			let vote = Self::community_vote_of(who, poll_index).ok_or(Error::<T>::NoVoteCasted)?;
 
+			let (vote, _) = CommunityVotes::<T>::get(poll_index, membership_id).ok_or(Error::<T>::NoVoteCasted)?;
 			let vote_multiplier = match CommunityDecisionMethod::<T>::get(community_id) {
 				DecisionMethod::Rank => T::MemberMgmt::rank_of(&community_id, &membership_id)
 					.unwrap_or_default()
 					.into(),
 				_ => 1,
 			};
+
 			let vote_weight = VoteWeight::from(&vote);
 			tally.remove_vote(vote.say(), vote_multiplier * vote_weight, vote_weight);
 
-			let reason = HoldReason::VoteCasted(poll_index).into();
-			CommunityVotes::<T>::remove(who, poll_index);
-
-			match vote {
-				Vote::AssetBalance(_, asset_id, amount) => {
-					T::Assets::release(asset_id.clone(), &reason, who, amount, Precision::BestEffort).map(|_| ())
-				}
-				Vote::NativeBalance(..) => T::Balances::thaw(&reason, who),
-				_ => Ok(()),
-			}
+			CommunityVotes::<T>::remove(poll_index, membership_id);
+			Self::update_locks(who)
 		})
 	}
 
-	fn do_lock_for_vote(who: &AccountIdOf<T>, poll_index: &PollIndexOf<T>, vote: &VoteOf<T>) -> DispatchResult {
-		let reason = HoldReason::VoteCasted(*poll_index).into();
-		CommunityVotes::<T>::insert(who, poll_index, vote.clone());
+	pub(crate) fn do_unlock(
+		who: &AccountIdOf<T>,
+		membership_id: MembershipIdOf<T>,
+		poll_index: PollIndexOf<T>,
+	) -> DispatchResult {
+		T::MemberMgmt::check_membership(who, &membership_id).ok_or(Error::<T>::NotAMember)?;
+		CommunityVotes::<T>::remove(poll_index, membership_id);
+		Self::update_locks(who)
+	}
+
+	pub(crate) fn update_locks(who: &AccountIdOf<T>) -> DispatchResult {
+		use sp_runtime::traits::Zero;
+		let reason = HoldReason::VoteCasted.into();
 
-		match vote {
-			Vote::AssetBalance(_, asset_id, amount) => T::Assets::hold(asset_id.clone(), &reason, who, *amount),
-			Vote::NativeBalance(_, amount) => {
-				T::Balances::set_frozen(&reason, who, *amount, frame_support::traits::tokens::Fortitude::Polite)
+		let mut assets_locked_amount: Vec<(AssetIdOf<T>, AssetBalanceOf<T>)> = vec![];
+		let mut native_locked_amount: NativeBalanceOf<T> = Zero::zero();
+
+		for (locked_vote, _) in CommunityVotes::<T>::iter_values().filter(|(_, voter)| voter == who) {
```

Comment:
Done. Now locks are stored on `CommunityVoteLocks`, while the tuple `(vote, voter)` is preserved to allow deleting the previous voter's account lock in case the new voter's account is not the same (i.e. membership has been transferred).